### ISSUE TITLE
NBT -> Component datafixer for updating existing 1.20 worlds to 1.21

### ISSUE THREAD

**Issue by Robotgiggle** [gamma-delta/HexMod#1214]
### Describe the feature

Currently, if you update a world that was using 1.20 hex to 1.21, anything stored in custom NBT tags (casting items, iota holders, phials, etc) will vanish. The data is actually all still there, since it gets moved into the `minecraft:custom_data` component when you update the world, but it's not accessible ingame since 1.21 hex expects all that stuff to be in its own custom components.

We should implement some kind of datafixer to check for the 1.20-era custom NBT in the custom_data component and move it into the appropriate 1.21-era components, so that people can update existing worlds without losing all their stored hexes and iotas.

**Comment by Olfi01** [comment]
Are you thinking something that is run every time an item is loaded or some kind of manually run one-time script?

**Comment by Robotgiggle** [comment]
If there's a way to hook into the vanilla world-updating process that turns item NBT into components (via mixin, maybe?) to do our stuff at the same time, that would be ideal. If that's not possible, though, a check every time items are loaded could also work.

**Comment by Olfi01** [comment]
After digging through the Minecraft source code a bit, I think the class responsible for this is likely ItemStackComponentizationFix, which is one of many Datafixes applied by Mojangs DataFixer. We might just be able to implement and register our own class extending DataFix which does the component fixing for us, but we would probably have to go through each and every component we register.
Another option would be to mixin right into that Datafix, but I would probably like to avoid that if possible.

**Comment by Robotgiggle** [comment]
Registering our own DataFix definitely seems like the best solution. Going through all our custom components might be a little tedious, but it makes sense that it would be necessary to do something like this.

**Comment by Robotgiggle** [comment]
Oh also, it looks like there was already an attempt to do this on the Fabric side [here](https://github.com/FallingColors/HexMod/blob/49432f373f18acdafc0d164500942328cd8eb9cf/Fabric/src/main/java/at/petrak/hexcasting/fabric/cc/HexCardinalComponents.java#L92), but it doesn't seem to have worked - I can confirm that opening a Fabric 1.20 world with Fabric 1.21 does indeed wipe all of hex's custom data.

I'm not sure _why_ it didn't work, but it seems to be trying to handle the migration via Cardinal Components. Doing it directly via vanilla systems might work better.

**Comment by Olfi01** [comment]
I've got most of the ItemStack fixing worked out [here](https://github.com/Olfi01/HexMod/tree/issue/1214-fixup-item-components), ~but I'm still struggling to figure out where the BlockEntity data is getting fixed in vanilla~
Placed Slate block entites also break because the pattern serialization has changed between 1.20 and 1.21

**Comment by Olfi01** [comment]
I opened #1225 to track that separately since I think my PR will already be quite large.
